### PR TITLE
[CALCITE-4642] Use TypeSystem from Config in Planner

### PR DESCRIPTION
**Pull request by njriasan**
Fixes a bug in planner where the typeSystem being used wasn't the one provided by the config. Added a test demonstrating the support on typeSystems that have different behavior on unioning chars.

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/tools/PlannerTest.java [hunk 1]
```diff
@@ -1525,4 +1527,65 @@ private void checkView(String sql, Matcher<String> matcher)
     final RelRoot root = planner.rel(validate);
     assertThat(toString(root.rel), matcher);
   }
+
+  /**
+   *  Helper function that checks that a Planner uses the correct type system when provided.
+   */
+  private void checkTypeSystem(String sql, RelDataTypeSystem typeSystem, String expectedRelExpr)
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/tools/PlannerTest.java:1534]: Small notes:

1. The naming does not reflect the actual implementation.
The better name would be `assertValidatedPlan`.
2. It is better place the method _after_ the test method, so the code is easier to read from top to bottom (higher-level methods shoud come before low-level ones, especially when lower-level methods have just one usage)

**Comment by njriasan** [comment]
Thanks for the feedback @vlsi. I made the suggested changes. Is there any additional change I need to make before I can run the workflows?

**Comment by vlsi** [comment]
Please use upper case for `[Calcite-4642]`, and please remove `fixed a bug` from the commit title/PR title.
In practice, `fixed a bug` adds little value, and it would be better to spend the letters explaining what is changed.

For instance: `Use TypeSystem from Config in Planner`

**Comment by njriasan** [comment]
> Please use upper case for `[Calcite-4642]`, and please remove `fixed a bug` from the commit title/PR title.
> In practice, `fixed a bug` adds little value, and it would be better to spend the letters explaining what is changed.
> 
> For instance: `Use TypeSystem from Config in Planner`

Done. Thank you for the feedback.

**Review by hsyuan — COMMENTED** [review]
@njriasan, please remove dot "." in the commit message.

**Comment by julianhyde** [comment]
I'd remove "Helper function that checks that", which is superfluous.

Change the last parameter from `String expectedRelExpr` to `Matcher<String> planMatcher` and people will start using the function as an assert.

Put the helper function before the class. We tend to put inner classes and interfaces last.

Anyway. These are very pedantic points. It's a nice contribution; thank you!

**Comment by njriasan** [comment]
Thank you for the feedback everyone. I have made the desired changes. Could someone approve running the workflows?

**Comment by njriasan** [comment]
It looks like all the workflows are passing. Are there any additional steps I need to take in order to get this PR merged?

**Comment by njriasan** [comment]
@vlsi what is the process for getting a PR merged?

File: core/src/test/java/org/apache/calcite/tools/PlannerTest.java [hunk 2]
```diff
@@ -1525,4 +1527,67 @@ private void checkView(String sql, Matcher<String> matcher)
...
+    final String expectedVarying = "LogicalProject("
+        + "EXPR$0=["
+        + "CASE(<>($0, 30),"
+        + " 'hi':VARCHAR(5), "
+        + "'world':VARCHAR(5))])\n"
+        + "  LogicalValues("
+        + "tuples=[[{ 10, 'Sales' },"
+        + " { 20, 'Marketing' },"
+        + " { 30, 'Engineering' },"
+        + " { 40, 'Empty' }]])\n";
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/tools/PlannerTest.java:1546]:
```suggestion
    final String expectedVarying = ""
        + "LogicalProject(EXPR$0=[CASE(<>($0, 30), 'hi':VARCHAR(5), 'world':VARCHAR(5))])\n"
        + "  LogicalValues(tuples=[[{ 10, 'Sales' }, { 20, 'Marketing' }, { 30, 'Engineering' }, { 40, 'Empty' }]])\n";
```

File: core/src/test/java/org/apache/calcite/tools/PlannerTest.java [hunk 3]
```diff
@@ -1525,4 +1527,67 @@ private void checkView(String sql, Matcher<String> matcher)
...
+    final String expectedDefault = "LogicalProject("
+        + "EXPR$0=["
+        + "CASE(<>($0, 30),"
+        + " 'hi   ', "
+        + "'world')])\n"
+        + "  LogicalValues("
+        + "tuples=[[{ 10, 'Sales      ' },"
+        + " { 20, 'Marketing  ' },"
+        + " { 30, 'Engineering' },"
+        + " { 40, 'Empty      ' }]])\n";
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/tools/PlannerTest.java:1556]:
```suggestion
    final String expectedDefault = ""
        + "LogicalProject(EXPR$0=[CASE(<>($0, 30), 'hi   ', 'world')])\n"
        + "  LogicalValues(tuples=[[{ 10, 'Sales      ' }, { 20, 'Marketing  ' }, { 30, 'Engineering' }, { 40, 'Empty      ' }]])\n";
```

**Review by vlsi — APPROVED** [review]
What do you think of avoiding extra linebreaks in the plans?

File: core/src/test/java/org/apache/calcite/tools/PlannerTest.java (hunk 2 above)
Comment (julianhyde) [reply core/src/test/java/org/apache/calcite/tools/PlannerTest.java:1546]: I'm in favor of inserting line-breaks by hand to keep lines short. I think the previous is fine.

Comment (vlsi) [reply core/src/test/java/org/apache/calcite/tools/PlannerTest.java:1546]: Ok, whatever. It took me a while to understand the plan, as I had to search for `\n`, so I asked.

Comment (julianhyde) [reply core/src/test/java/org/apache/calcite/tools/PlannerTest.java:1546]: Yeah, that's valid. It's difficult to see the  relational operators 'LogicalProject' and 'LogicalValues'. Sometimes I ensure that continuation lines have no preceding spaces at, which visually distinguishes them from the indented relational operators. Like this:

```
"LogicalProject(EXPR$0=[CASE(<>($0, 30), "
        + "'hi   ', "
        + "'world')])\n"
        + "  LogicalValues("
        + "tuples=[[{ 10, 'Sales      ' }, "
        + "{ 20, 'Marketing  ' }, "
        + "{ 30, 'Engineering' }, "
        + "{ 40, 'Empty      ' }]])\n";
```

It works well in large, very-indented plans.

Comment (vlsi) [reply core/src/test/java/org/apache/calcite/tools/PlannerTest.java:1546]: Exactly. I do see the reason to wrap the plan line after 200-300 chars, however, the current case is not worth the discussion.
I think non-wrapped is readable, and the wraps make it harder to read.

However, I do not object, and the rest in the patch looks great.

File: core/src/test/java/org/apache/calcite/tools/PlannerTest.java (hunk 3 above)
Comment (njriasan) [reply core/src/test/java/org/apache/calcite/tools/PlannerTest.java:1556]: Thanks for the suggestion. I'll merge this change to make it more readable.

**Review by NobiGo — APPROVED** [review]
LGTM

**Comment by njriasan** [comment]
Can we merge this?

**Pull request closed without merging**